include/libdnf/advisory/advisory_query.hpp

```diff
@@ -0,0 +1,114 @@
+
+#include "libdnf/advisory/advisory.hpp"
+#include "libdnf/advisory/advisory_collection.hpp"
+#include "libdnf/advisory/advisory_package.hpp"
+#include "libdnf/advisory/advisory_reference.hpp"
+#include "libdnf/advisory/advisory_sack.hpp"
+#include "libdnf/common/sack/query_cmp.hpp"
+#include "libdnf/rpm/package.hpp"
+
+namespace libdnf::advisory {
+
+class AdvisoryQuery {
+public:
+    explicit AdvisoryQuery(AdvisorySack * sack);
+    AdvisoryQuery(const AdvisoryQuery & src);
+    AdvisoryQuery(AdvisoryQuery && src);
+    ~AdvisoryQuery();
+
+    AdvisoryQuery & operator=(const AdvisoryQuery & src);
+    AdvisoryQuery & operator=(AdvisoryQuery && src) noexcept;
+
+    struct NotSupportedCmpType : public RuntimeError {
+        using RuntimeError::RuntimeError;
+        const char * get_domain_name() const noexcept override { return "libdnf::advisory::AdvisoryQuery"; }
+        const char * get_name() const noexcept override { return "NotSupportedCmpType"; }
+        const char * get_description() const noexcept override { return "AdvisoryQuery exception"; }
+    };
+
+    /// Filter Advisories by name
+    ///
+    /// @param pattern      Pattern used when matching agains advisory names.
+    /// @param cmp_type     What comparator to use with pattern, allows: EQ, GLOB, IGLOB.
+    AdvisoryQuery & ifilter_name(const std::string & pattern, sack::QueryCmp cmp_type = libdnf::sack::QueryCmp::EQ);
+    AdvisoryQuery & ifilter_name(
+        const std::vector<std::string> & patterns, sack::QueryCmp cmp_type = libdnf::sack::QueryCmp::EQ);
+
+    AdvisoryQuery & ifilter_type(const Advisory::Type type, sack::QueryCmp cmp_type = libdnf::sack::QueryCmp::EQ);
+    AdvisoryQuery & ifilter_type(
+        const std::vector<Advisory::Type> & types, sack::QueryCmp cmp_type = libdnf::sack::QueryCmp::EQ);
+
+    AdvisoryQuery & ifilter_reference(
+        const std::vector<std::string> & reference_id,
+        std::vector<AdvisoryReferenceType> types,
+        sack::QueryCmp cmp_type = libdnf::sack::QueryCmp::EQ);
+    //TODO(amatej): What about other reference fitlers (title, url?) - do we want them?
+    AdvisoryQuery & ifilter_CVE(const std::string & pattern, sack::QueryCmp cmp_type = libdnf::sack::QueryCmp::EQ);
+    AdvisoryQuery & ifilter_CVE(
+        const std::vector<std::string> & patterns, sack::QueryCmp cmp_type = libdnf::sack::QueryCmp::EQ);
+    AdvisoryQuery & ifilter_bug(const std::string & pattern, sack::QueryCmp cmp_type = libdnf::sack::QueryCmp::EQ);
+    AdvisoryQuery & ifilter_bug(
+        const std::vector<std::string> & patterns, sack::QueryCmp cmp_type = libdnf::sack::QueryCmp::EQ);
+
+    AdvisoryQuery & ifilter_severity(const std::string & pattern, sack::QueryCmp cmp_type = libdnf::sack::QueryCmp::EQ);
+    AdvisoryQuery & ifilter_severity(
+        const std::vector<std::string> & patterns, sack::QueryCmp cmp_type = libdnf::sack::QueryCmp::EQ);
+
+    AdvisoryQuery & ifilter_applicability(const bool value = true);
```

Comment:
I am not sure about the name of the filter or value argument.

Reply:
I am also not sure about following filters: `ifilter_applicability`, `ifilter_package`, `ifilter_package`. I believe that the consequence of their presence will be misleading to end users. I suggest that the operations performed by the filters must be perform on AdvisoryPackage or AdvisoryCollection level. May I ask you to add TODO? Replacement of these method can be `get_advisory_packages` that will return AdvisoryPackageSet with method to remove packages from not applicable collections.

Reply:
I have removed `ifilter_applicability` and `ifilter_package`. The only method kept is `ifilter_packages` since it is implementing most of the logic, but I have added a TODO that we might want to remove it if it ends up not used.